### ISSUE TITLE
An alternative solution for hGetContent error on Windows when non-English

### ISSUE THREAD

**Issue by ice1000** [agda/agda#4161]
On non-English Windows, people may encounter some stupid issues of Haskell, like the following:

```
hGetContents: invalid argument (invalid byte sequence)
```

This issue is solved and documented here: https://wiki.portal.chalmers.se/agda/pmwiki.php?n=Main.WindowsNotes-2-4-2 and the solution points you to a link that no longer exists.
Good news: I have found an alternative solution. You actually don't need to change your system encoding, you can simply do

```
CHCP 65001
```

and then install Agda via `stack install` or `make install` or `cabal new-install` or whatever.

I wish this can be documented somewhere in agda.readthedocs.io, as Agda wiki is really really hard for people to navigate to.

**Comment by ice1000** [comment]
https://github.com/chrisdone/intero/issues/603#issuecomment-453784278

**Comment by nad** [comment]
> This issue is solved and documented here: https://wiki.portal.chalmers.se/agda/pmwiki.php?n=Main.WindowsNotes-2-4-2

The cause of the problem discussed there is presumably that cpphs reads (or used to read?) Haskell files using the currently active code page instead of UTF-8 (which GHC uses, more or less). Was your problem also caused by cpphs, or something else?

**Comment by ice1000** [comment]
> The cause of the problem discussed there is presumably that cpphs reads (or used to read?) Haskell files using the currently active code page instead of UTF-8 (which GHC uses, more or less). Was your problem also caused by cpphs, or something else?

It's Happy this time.

**Comment by nad** [comment]
In that case I suggest that you report this bug to the Happy developers.

**Comment by ice1000** [comment]
> In that case I suggest that you report this bug to the Happy developers.

This is unlikely a Happy issue, I think GHC should be blamed. It happens to [other Haskell projects](https://github.com/chrisdone/intero/issues/603#issuecomment-453784278) (also [HIE](https://github.com/haskell/haskell-ide-engine/issues/667), [hakyll](https://github.com/jaspervdj/hakyll/issues/593), [another Haskell proj](https://github.com/koalaman/shellcheck/issues/324)) as well.

(or maybe it's just Happy didn't do error handling? I'm not sure)

**Comment by nad** [comment]
If standard functions like `readFile` are used, then the default character encoding used when reading files [depends on the locale/code page](https://hackage.haskell.org/package/base-4.12.0.0/docs/System-IO.html#g:23). If you want to use a specific character encoding, then you should ask for it.

Here is the function that Agda uses to read text files which are supposed to use the UTF-8 character encoding:

https://github.com/agda/agda/blob/9afeaa33b610a1f2ab33921a8fdf36c491c211e1/src/full/Agda/Utils/IO/UTF8.hs#L33-L43

It seems as if Happy [uses `readFile`](https://github.com/simonmar/happy/blob/27596ff0ce0171d485bf96d38943ffc760923c90/src/Main.lhs#L72-L74).

**Comment by ice1000** [comment]
Wow, I just see an Agda 2.6.0.1 mirror: http://homepage.divms.uiowa.edu/~astump/agda/

**Comment by banacorn** [comment]
For posterity: if `CHCP 65001` does not work, please consider [changing the system locale](https://superuser.com/a/1435645) instead. 

My guess is that `CHCP 65001` applies only to child processes but not grandchild processes.

**Comment by andreasabel** [comment]
At least for building Agda, this should be fixed (#5465) by de-unicoding `Parser.y`.
